Mergeable rustdoc cross-crate info (#3662)
rust-lang/rfcs: text/0000-mergeable-rustdoc-cross-crate-info.md

- Feature Name: `mergable_rustdoc_cross_crate_info`
- Start Date: 2024-06-18
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary

> Mergeable cross-crate information in rustdoc. Facilitates the generation of documentation indexes in workspaces with many crates by allowing each crate to write to an independent output directory. Final documentation is rendered with a lightweight merge step. Configurable with command-line flags, this proposal writes a `doc.parts` directory to hold pre-merge cross-crate information. Currently, rustdoc requires global mutable access to a single output directory to generate cross-crate information, which is an obstacle to integrating rustdoc in build systems that enforce the independence of build actions.

Reviewer: Editorial: These aren't full sentences, and it's kind-of hard to understand what's going on.

# Motivation

The main goal of this proposal is to facilitate users producing a documentation bundle of every crate in a large environment. When a crate needs to be re-documented, only a relatively lightweight merge step will be needed to produce an updated documentation bundle. This proposal is to facilitate the creation and updating of these bundles.

This proposal also targets documenting individual crates and their dependencies in non-cargo build systems. As will be explained, doc targets in non-cargo build systems often do not support cross-crate information.

There are some files in the rustdoc output directory that are read and overwritten during every invocation of rustdoc. This proposal refers to these files as **cross-crate information**, or **CCI**, as in <https://rustc-dev-guide.rust-lang.org/rustdoc.html#multiple-runs-same-output-directory>.

Build systems may run build actions in a distributed environment across separate logical filesystems. It might also be desirable to run rustdoc in a lock-free parallel mode, where every rustdoc process writes to a disjoint set of files.

> Cargo fully supports cross-crate information, at the cost of requiring global read-write access to the doc root (`target/doc`). There are significant scalability issues with this approach.

Reviewer: Editorial: It's worth clarifying that this isn't just a property of cargo, but they way cargo integrates with rustdoc.

These considerations motivate adding an option for outputting partial CCI (parts), which are merged (linked) with a later step.

This RFC has the goal of enabling the future deprecation of the default (called `--merge=shared` here) practice of appending to cross-crate information files in the doc root.

# Guide-level explanation

## New flag summary

More details are in the Reference-level explanation.

* `--parts-out-dir=path/to/doc.parts/<crate-name>`: Write cross-crate linking information to the given directory (only usable with the `--merge=none` mode). This information allows linking the current crate's documentation with other documentation at a later rustdoc invocation.
* `--include-parts-dir=path/to/doc.parts/<crate-name>`: Include cross-crate information from this previously written `doc.parts` directories into a collection that will be written by the current invocation of rustdoc. May only be provided with `--merge=finalize`. May be provided any number of times.

> * `--merge=none`: Do not write cross-crate information to the `--out-dir`. The flag `--parts-out-dir` may instead be provided with the destination of the current crate's cross-crate information parts.

Reviewer: Editorial: Given that the `--merge` flag is the main one controlling rustdoc's operation, it makes sense for it to be listed first, and then the supporting flags to be documented under the `--merge` values they may be used with.

> * `--merge=shared` (default): Append information from the current crate to any info files found in the `--out-dir`.

Reviewer 2: Since we plan to deprecate `--merge=shared` anyway, what do you think about not adding it in the first place? `--merge` already has to be an optional flag to avoid breaking changes.

Reviewer 2: Fair enough. I guess we won't ever be able to remove the shared implementation anyway, so it doesn't matter. And if we somehow did find a way to remove it while maintaining the outward behavior, that wouldn't be a breaking change anyway since it's an implementation detail.

Reviewer 3: Why do you say we won't ever be able to remove the shared implementation? I've been assuming that after updating Cargo and docs.rs, and providing a decently long deprecation period we could remove it.

Reviewer 2: Hmm, I'm trying to remember what I was thinking ^^. I think I was worried that there might be existing build systems or other tools that were depending on the shared representation, but we never guaranteed stability of it. So perhaps it would be fine to remove after all.

### Document crates, delaying generation of cross-crate information: `--merge=none`

Document crates using a dedicated HTML output directory and a dedicated "parts" output directory. No cross-crate data nor rendered HTML output is included from other crates.

This mode only renders the HTML item documentation for the current crate. It does not produce a search index, cross-crate trait implementations, or an index page. It is expected that users follow this mode with 'Link documentation' if these cross-crate features are desired.

In this mode, a user will provide `--parts-out-dir=<path to crate-specific directory>` and `--merge=none` to each crate's rustdoc invocation. The user should provide `--extern-html-root-url`, and specify a absolute final destination for the docs, as a URL. The `--extern-html-root-url` flag should be provided for each crate's rustdoc invocation, for every dependency.

> The same `--out-dir` may be used for multiple parallel rustdoc invocations, as rustdoc will continue to acquire an flock on the `--out-dir` to address conflicts. A user may select a different `--out-dir` for each crate's rustdoc invocation. 

Reviewer: Clarification: Why support this use-case? Is there any time that a user may want to have multiple rustdoc invocations with the same `--out-dir` while also having `--merge=none`

## Merge step

This proposal is capable of addressing two primary use cases. It allows developers to enable CCI in these scenarios:
* Documenting a crate and its transitive dependencies in parallel in build systems that require build actions to be independent
* Producing a documentation index of every crate in a workspace, in such a way that if one crate is updated, only the updated crates and an index have to be redocumented. This scenario is demonstrated in the Guide-level explanation.

CCI is not automatically enabled in either situation. A combination of the `--include-parts-dir`, `--merge`, and `--parts-out-dir` flags are needed to produce this behavior. This RFC provides a minimal set of tools that allow developers of build systems, like Bazel and Buck2, to create rules for these scenarios. 

> With separate `--out-dir`s, copying item docs to an output destination is needed. Rustdoc will never support the entire breadth of workflows needed to merge arbitrary directories, and will rely on users to run external commands like `mv`, `cp`, `rsync`, `scp`, etc. for these purposes. Most users are expected to use a single `--out-dir` for all crates, in which case these external tools are not needed.

Reviewer: Clairification: Is the idea here that tools that build systems that need to be hermetic will use seperate `--out-dir`s for there `--merge=none` steps and have to implement merging directorys themself, while cargo will have use the same `--out-dir` for each rustdoc invocation even after it stops using `--merge=shared` (even implicitly)

## Compatibility

This RFC does not alter previous compatibility guarantees made about the output of rustdoc. In particular it does not stabilize the presence of the rendered cross-crate information files, their content, or the HTML generated by rustdoc.

In the same way that the [rustdoc HTML output is unstable](https://rust-lang.github.io/rfcs/2963-rustdoc-json.html#:~:text=The%20HTML%20output%20of%20rustdoc,into%20a%20different%20format%20impractical), the content of `doc.parts` will be considered unstable. Between versions of rustdoc, breaking changes to the content of `doc.parts` should be expected. Only the presence of a `doc.parts` directory is promised, under `--parts-out-dir`. Merging cross-crate information generated by disparate versions of rustdoc is not supported. To detect whether `doc.parts` is compatible, rustdoc includes a version number in these files (see New directory: `doc.parts`).

> The implementation of the RFC itself is designed to produce only minimal changes to cross-crate info files and the HTML output of rustdoc. Exhaustively, the implementation is allowed to 

Reviewer: Editorial: What is this talking about? There was a linked draft implementation in the PR, but that was on the master branch of your fork, which now doesn't have it. It's worth clairifying that this is your WIP implementation (unless it isn't), and linking to it.

Author: I removed the reference. It was to a different RFC (rust-lang/rfcs#2963), but I don't think it is relevant anymore.

The limitation of `--extern-html-root-url` is that it needs to be provided with an absolute URL for the final docs destination. If your docs are hosted on `https://example.com/docs/`, this URL must be *known at documentation time*, and provided through `--extern-html-root-url=<crate name>=https://example.com/docs/`. *Absolute URLs*, instead of relative URLs, are generated for items in externally located crates. A future proposal may address this limitation by providing a command line option that generates relative URLs (like is done between items in the current crate, or other locally documented crates) for selected external crates, assuming that these crates will end up in the same bundle. The existing `--extern-html-root-url` is sufficient for the use cases envisioned by this RFC, despite the limitation.

# Drawbacks

> The WIP may change the sorting order of the elements in the CCI. It does not change the content of the documentation, and is intended to work without modifying Cargo and docs.rs.

Reviewer 2: ```suggestion
The implementation may change the sorting order of the elements in the CCI. It does not change the content of the documentation, and is intended to work without modifying Cargo and docs.rs.
```

## Buck2

The Buck2 build system has rules for building and testing rust binaries and libraries. <https://buck2.build/docs/prelude/globals/#rust_library>

It has a subtarget, `[doc]`, for generating rustdoc for a crate.

You can provide `extern-html-root-url`. You can document all crates independently and manually merge them but no cross-crate information would be shared.

buck2 does not natively merge rustdoc from separate targets. The buck2 maintainers have a [proprietary search backend](https://rust-lang.zulipchat.com/#narrow/stream/266220-t-rustdoc/topic/mergable.20rustdoc.20proposal/near/445952204) that merges and parses `search-index.js` files from separately documented crates. Their proprietary tooling does not handle cross-crate trait implementations from upstream crates. By implementing this merging directly in rustdoc, we could avoid fragmentation and bring cross-crate information to more consumers.

<https://github.com/facebook/buck2/blob/d390e31632d3b4a726aaa2aaf3c9f1f63935e069/prelude/rust/build.bzl#L213>

## Ninja [(GN)](https://fuchsia.dev/fuchsia-src/development/build/build_system/intro) + Fuchsia

> Currently, the Fuchsia project runs rustdoc on all of their crates to generate a [documentation index](https://fuchsia-docs.firebaseapp.com/rust/rustdoc_index/). This index is effectively generated as an [atomic step](https://cs.opensource.google/fuchsia/fuchsia/+/main:tools/devshell/contrib/lib/rust/rustdoc.py) in the build system. It takes [3 hours](https://ci.chromium.org/ui/p/fuchsia/builders/global.ci/firebase-docs/b8744777376580022225/overview) to document the ~2700 crates in the environment. With this proposal, building each crate's documentation could be done as separate build actions, which would have a number of benefits. These include parallelism, caching (avoid rebuilding docs unnecessarily), and robustness (automatically reject pull requests that break documentation).

Reviewer: Nit: It's worth making this a permalink to a specific commit, as if this does get merged I suspect this script will be significantly rewritten,

## New flag: `--merge=none|shared|finalize`

This flag corresponds to the three modes of invoking rustdoc described in 'Suggested workflows'. It controls two internal paramaters: `read_rendered_cci`, and `write_rendered_cci`.

When `write_rendered_cci` is active, rustdoc will output the rendered parts to the doc root (`--out-dir`). Rustdoc will generate files like `doc/search-index.js`, `doc/search.desc`, `doc/index.html`, etc if and only if this parameter is true.

When `read_rendered_cci` is active, rustdoc will look in the `--out-dir` for rendered cross-crate info files. These files will be used as the base. Any new parts that rustdoc generates with its current invocation and any parts fetched with `include-parts-dir` will be appended to these base files. When it is disabled, the cross-crate info files start empty and are populated with the current crate's info and any crates fetched with `--include-parts-dir`.

> * `--merge=shared` (`read_rendered_cci && write_rendered_cci`) is the default, and reflects the current behavior of rustdoc.

Reviewer 4: nit: also probably explain what the behavior is here, this is the main place the flags are documented

There are some files in the rustdoc output directory that are read and overwritten during every invocation of rustdoc. This proposal refers to these files as **cross-crate information**, or **CCI**, as in <https://rustc-dev-guide.rust-lang.org/rustdoc.html#multiple-runs-same-output-directory>.

Build systems may run build actions in a distributed environment across separate logical filesystems. It might also be desirable to run rustdoc in a lock-free parallel mode, where every rustdoc process writes to a disjoint set of files.

Cargo fully supports cross-crate information, at the cost of requiring global read-write access to the doc root (`target/doc`). There are significant scalability issues with this approach.

Rustdoc needing global mutable access to the files that encode this cross-crate information has implications for caching, reproducible builds, and content hashing. By adding an option to avoid this mutation, rustdoc will serve as a first-class citizen in non-cargo build systems.

These considerations motivate adding an option for outputting partial CCI (parts), which are merged (linked) with a later step.

# Guide-level explanation

> 

Author: I see the users of this feature as people who are orchestrating complex rustdoc builds, so my intention was to make this section useful to them. I agree it is more advanced / internal, but I would welcome suggestions about how to make it more digestible to these users or to a more general audience!

```shell
mkdir -p t/src s/src i/src merged/doc
echo "pub trait T {}" > t/src/lib.rs
echo "pub struct S; impl t::T for S {}" > s/src/lib.rs
MERGED=file://$(realpath merged/doc)
```

[Actively use](https://doc.rust-lang.org/rustc/command-line-arguments.html#--extern-specify-where-an-external-library-is-located) `t` and `s` in `i`. The `extern crate` declarations are not needed if the crates are otherwise referenced in the index; intra-doc links are enough.

```shell
echo "extern crate t; extern crate s;" > i/src/lib.rs
```

Compile the crates.

```shell
rustc --crate-name=t --crate-type=lib --edition=2021 --emit=metadata --out-dir=t/target t/src/lib.rs
rustc --crate-name=s --crate-type=lib --edition=2021 --emit=metadata --out-dir=s/target --extern t=t/target/libt.rmeta s/src/lib.rs
```

Document `s` and `t` independently, providing `--merge=none`, `--write-info-json`.

```shell
rustdoc \

>     -Z unstable-options \

Reviewer 2: Just to be 100% clear, as this example implies, all of these flags will initially be gated behind `-Z unstable-options`, correct?

Author: I think so. The example scraping RFC #3123 landed flags that were behind -Zunstable-options, even though the RFC didn't mention that they would be unstable.

    --crate-name=t \
    --crate-type=lib \
    --edition=2021 \
    --out-dir=t/target/doc \
    --extern-html-root-url t=$MERGED \
    --merge=none \

>     --write-info-json=t/target/doc.parts/t/crate-info.json \

Reviewer 2: I wonder if we can come up with a more descriptive name than `doc.parts`? Maybe `doc.cache`, `doc.shared`, `doc.global`, or `doc.metadata`? IIUC, the name would only be hardcoded in the cargo implementation, but it'd still be good to have a clear name; also we should explicitly decide whether the name and structure of the directory would have any stability.

Author: Yeah, the implementation itself would be agnostic to the name of the directory. I think the structure of the directory should not be stable, in the same way that we agreed that the contents of crate-info.json should not bestable. `doc.parts` is meant to invoke the fact that it includes pieces / parts of files that are merged together. It also invokes the .part suffix for downloaded files, which communicates that `doc.parts` will not be readable until it's fully merged.

├── s
│   ├── src
│   │   └── lib.rs
│   └── target
│       ├── doc
│       │   ├── help.html
│       │   ├── .lock
│       │   ├── s
│       │   │   ├── all.html
│       │   │   ├── index.html
│       │   │   ├── sidebar-items.js
│       │   │   └── struct.S.html
│       │   ├── search.desc
│       │   │   └── s
│       │   │       └── s-desc-0-.js
│       │   ├── settings.html
│       │   └── src
│       │       └── s
│       │           └── lib.rs.html
│       ├── doc.parts
│       │   └── s
│       │       └── crate-info.json
│       └── libs.rmeta
└── t
    ├── src
    │   └── lib.rs
    └── target
        ├── doc
        │   ├── help.html
        │   ├── .lock
        │   ├── search.desc
        │   │   └── t
        │   │       └── t-desc-0-.js
        │   ├── settings.html
        │   ├── src
        │   │   └── t
        │   │       └── lib.rs.html
        │   └── t
        │       ├── all.html
        │       ├── index.html
        │       ├── sidebar-items.js
        │       └── trait.T.html
        ├── doc.parts
        │   └── t
        │       └── crate-info.json
        └── libt.rmeta
</pre>

</details>

## Suggested workflows

### No cross-crate information 

> Provide `--merge=none` to every invocation of rustdoc.

Reviewer 5: Now that I'm thinking about it: couldn't it be the default so there is no need to provide it all the time? Therefore we could remove the `none` value and just expect that if `--merge` isn't provided, then this option is disabled.

Author: `--merge=read-write` should be implied if no `--merge` option is provided, because `--merge=read-write` is designed to match the existing behavior of rustdoc. Cargo, the rustdoc test suite, and other tools that use rustdoc expect to be able to run rustdoc with a shared output directory, and get cross-crate information. If `--merge=none` was made the default, these tools would lose cross-crate info, unless they are patched to use `--merge=read-write`.

Reviewer 5: Arf, sorry yes you're right. Please ignore this comment. ^^'

### Cross-crate information, mutate shared directory

> Use `--merge=read-write`, and specify the same `--out-dir` to every invocation of rustdoc. `--merge=read-write` will be the default value if `--merge` is not provided. This is the workflow that Cargo uses.

Reviewer 5: Another thought I had: why not provide `read` and `write` values and possibly allow them to be separated by a comma like `--merge=read,write` or `--merge=write,read`. Like that, only two values. :)

Reviewer 5: Hum, since `read` by itself is useless, not sure if it's a good idea or not...

Author: I think the command line help text is easier to write for `--merge=read-write|write-only|none`, since we don't have to explain that there are useless states.

Reviewer 5: Yes I don't really see the need for it.

# Reference-level explanation

The existing cross-crate information files, like `search-index.js`, all are lists of elements, rendered in an specified way (e.g. as a JavaScript file with a JSON array or an HTML index page containing an unordered list). The current rustdoc (in `write_shared`) pushes the current crate's version of the CCI into the one that is already found in `doc`, and renders a new version. The rest of the proposal uses the term **part** to refer to the pre-merged, pre-rendered element of the CCI. This proposal does not add any new CCI or change their contents (modulo sorting order, whitespace).

## New file: `crate-info.json`

`crate-info.json` is an artifact that encodes the partial contents and destination of several cross-crate information files. It only encodes information about a single-crate. This file is written if `--write-info-json` is provided. The current crate's information and any `crate-info.json` added through `--include-info-json` are merged and rendered if `--merge=read-write` or `--merge=write-only` are provided.

> The content of `crate-info.json` is unstable. Rustdoc only guarantees that it will accept `crate-info.json` files written by the same version of rustdoc, and rustdoc is the only explicitly supported consumer of `crate-info.json`. Only the presence of `crate-info.json` is stabilized. Non-normatively, there are several pieces of information that `crate-info.json` may contain:

Reviewer 2: Do we strictly *need* to stabilize even the presence of `crate-info.json`? It seems like we should only stabilize the existence of *some* "parts" directory whose contents are opaque and the fact that rustdoc will accept a "parts" directory produce by the same version of rustdoc. In other words, bring it up a level and only specify/stabilize the directory rather than the (currently sole, but in the future maybe one of several) file within it.

Author: I agree with you, and I will make this change.

Author: That's an important concern! I will mention that, and the fact that the directory structure may change when the edition changes.

Reviewer 2: > We kind of do, most custom build systems need to explicitly declare which files are intended to "escape containment".

Hmm, is it not possible to specify that an entire directory escapes containment? Or use a glob within that directory?

If those are not possible, then I would still prefer not to require an edition change for the directory's structure to change.

Author: I removed the edition requirement and said that rustdoc will make an effort to use a crate-info file, but the contents are formally unstable.

It does not document crates' dependencies. `search-index.js`, for example, is both a dependency and an output file for rustdoc in multi-crate documentation workspaces. If it is declared as a dependency in this way, Bazel could not build docs for the members of an environment in parallel with a single output directory, as it strictly enforces hermiticity. For a recursive, parallel rustdoc to ever serve as a first-class citizen in Bazel, changes similar to the ones described in this proposal would be needed.

There is an [open issue](https://github.com/bazelbuild/rules_rust/issues/1837) raised about the fact that Bazel does not document crates dependencies. The comments in the issue discuss a pull request on Bazel that documents each crates dependencies in a separate output directory. It is noted in the discussion that this solution, being implemented under the current rustdoc, "doesn't scale well and it should be implemented in a different manner long term." In order to get CCI in a mode like this, rustdoc would need to adopt changes, like the ones in this proposal, for merging cross-crate information.

## Buck2

> The Buck2 build system has rules for building and testing rust binaries and libraries. <https://buck2.build/docs/api/rules/#rust_library>

Reviewer 2: nit, but this link seems to be broken -- did you mean this instead?

```suggestion
The Buck2 build system has rules for building and testing rust binaries and libraries. <https://buck2.build/docs/prelude/globals/#rust_library>
```

Author: Yup, that looks like the new link! Thank you!

The proposition, to allow users of rustdoc the flexibility of not having to produce an index, is to allow rustdoc to be run in a mode where no target crate is provided. It would generate rendered cross-crate information based only on what is provided through `--include-info-json`. The source crates are provided to rustdoc, through a mechanism like `--extern`, rustdoc merges and writes the CCI, and copies the item and module links to the doc root. This would require more extensive changes, as rustdoc assumes that it is invoked with a target crate. This mode is somewhat analogous to the [example scraping mode](https://github.com/rust-lang/rfcs/blob/master/text/3123-rustdoc-scrape-examples.md). Having to create an index crate, that actively uses all of the crates in the environment, might prohibit the use of this feature in settings where users do not intend to produce an index, or where exhaustively listing all dependencies (to `--extern` them) is difficult.

## Unconditionally generating the `doc.parts` files?

> Generate no extra files (current) vs. unconditionally creating `doc.parts` to enable more complex future CCI (should consider).

Reviewer 2: We should discuss this more.

Author: I am pretty strongly in favor of maintaining the default behavior of rustdoc, with no extra parts files, because that localizes the effort of the implementation entirely to rustdoc itself. I think needing to make changes to external tools would complicate the landing and implementation of this PR, and opens the door for too many potential issues.

Reviewer 2: Fair enough :+1:

* Reading exclusively from `crate-info.json` is simpler than the existing `serde_json` dependency for extracting the blobs, as opposed to handwritten CCI-type specific parsing (current)
* With this proposal, there will be duplicate logic to read from both `doc.parts` files and rendered CCI.

[@jsha proposes](https://github.com/rust-lang/rfcs/pull/3662#issuecomment-2184077829) unconditionally generating and reading from `crate-info.json`, with no appending to the rendered crate info.

## Item links?

Require users to pass `--extern-html-root-url` on all external dependencies (current), vs. add a new flag to facilitate missing docs links being generated across all user crates (should consider)

Cross-crate links are an important consideration when merging documentation. rustdoc provides a flag, `--extern-html-root-url`, which can provide fine-grain control over the generated URL for documentation. We believe this is already sufficient for generating cross-crate links, and we will make no changes to the generation of item links. Example usage of this flag is in the Guide-level explanation.

> A proposal is to provide more options to facilitate cross-crate links. For example, we could add a flag that implies `--extern-html-root-url` and `--include-info-json` on all crates passed through `--extern`. It could be something like `--extern-html-root-url-all-user-crates`. User crates would be taken to mean the transitive dependencies of the current crate, excluding the standard library. This is complicated by things like <https://doc.rust-lang.org/cargo/reference/unstable.html#build-std>.

Reviewer 2: Could you explain more why `--extern-html-root-url` is needed/recommended in more situations after the implementation of this proposal? It's a bit unclear to me. Also, by "external crates" do you mean crates that do not belong to the current workspace?

Author: Just added this to the RFC:
For the purpose of generating cross-crate links, rustdoc currently classifies the location of crates as external, local, or unknown (relative to the crate in the current invocation of rustdoc). Local crates share the `--out-dir`. with the current crate. External crates have documentation that could not be found in the current `--out-dir`, but otherwise have a known location. Item links are not generated to crates with an unknown location. When the `--extern-html-root-url=<crate name>=<url>` flag is provided, an otherwise unknown crate <crate name> becomes an externally located crate, forcing it to generate item links.

This is of relevance to this proposal, because users who document crates with separate `--out-dir`s may still expect cross-crate links to work. Currently, `--extern-html-root-url` is the exclusive mechanism for specifying link destinations for crates who would otherwise have an unknown location. We will expect users to provide `--extern-html-root-url` for all direct dependencies of a crate they are documenting. Example usage of this flag is in the Guide-level explanation.

Reviewer 2: Excellent, thanks!

## Reuse existing option?

Create a new flag, `--merge` (proposed), vs. use existing option `no_emit_shared`.

There is a render option, `no_emit_shared`, which is used to conditionally generate the cross-crate information. It also controls the generation of static files, user CSS, etc.

This option is not configurable from the command line, and appears to be enabled unless rustdoc is run in its example scraping mode.

We could make it configurable from the command line, unconditionally generate `doc.parts`, and use it to gate the merging of CCI.

We could also make it configurable from the command line, and use it to gate the generation of `doc.parts` and the generation of all of the shared files.

We could also leave it as-is: always false unless we're scraping examples, and gate the generation of `doc.parts` and the generation of all of the shared files.

# Future possibilities

> This change could begin to facilitate type aliases and the trait aliases being

Author: Ah, so it would not be desirable to bake trait impls on type aliases into the HTML. I will remove that as a future possibility.

There are some files in the rustdoc output directory that are read and overwritten during every invocation of rustdoc. This proposal refers to these files as **cross-crate information**, or **CCI**, as in <https://rustc-dev-guide.rust-lang.org/rustdoc.html#multiple-runs-same-output-directory>.

Build systems may run build actions in a distributed environment across separate logical filesystems. It might also be desirable to run rustdoc in a lock-free parallel mode, where every rustdoc process writes to a disjoint set of files.

Cargo fully supports cross-crate information, at the cost of requiring global read-write access to the doc root (`target/doc`). There are significant scalability issues with this approach.

Rustdoc needing global mutable access to the files that encode this cross-crate information has implications for caching, reproducible builds, and content hashing. By adding an option to avoid this mutation, rustdoc will serve as a first-class citizen in non-cargo build systems.

These considerations motivate adding an option for outputting partial CCI (parts), which are merged (linked) with a later step.

# Guide-level explanation

> In this example, there is a crate `t` which defines a trait `T`, and a crate `s` which defines a struct `S` that implements `T`. Our goal in this demo is for `S` to appear as in implementer in `T`'s docs, even if `s` and `t` are documented independently. This guide will be assuming that we want a crate `i` that serves as our documentation index. See the Unresolved questions section for ideas that do not require an index crate.

Reviewer 5: ```suggestion
In this example, there is a crate `t` which defines a trait `T`, and a crate `s` which defines a struct `S` that implements `T`. Our goal in this demo is for `S` to appear as an implementer in `T`'s docs, even if `s` and `t` are documented independently. This guide will be assuming that we want a crate `i` that serves as our documentation index. See the Unresolved questions section for ideas that do not require an index crate.
```

Document `s` and `t` independently, providing `--merge=none` and `--write-info-json=<crate name>/target/doc.parts`

```shell
rustdoc -Z unstable-options --crate-name=t --crate-type=lib --edition=2021 --out-dir=t/target/doc --extern-html-root-url t=$(MERGED) --merge=none --write-info-json=t/target/doc.parts t/src/lib.rs
rustdoc -Z unstable-options --crate-name=s --crate-type=lib --edition=2021 --out-dir=s/target/doc --extern-html-root-url s=$(MERGED) --extern-html-root-url t=$(MERGED) --merge=none --write-info-json=t/target/doc.parts --extern t=t/target/libt.rmeta s/src/lib.rs
```

Link everything with a final invocation of rustdoc on `i`. We will provide `--merge=write-only` and `--include-info-json=<path>`. See the Reference-level explanation about these flags.

```shell
rustdoc -Z unstable-options --crate-name=i --crate-type=lib --edition=2021 --enable-index-page --out-dir=i/target/doc --extern-html-root-url s=$(MERGED) --extern-html-root-url t=$(MERGED) --extern-html-root-url i=$(MERGED) --merge=write-only --include-info-json t=t/target/doc.parts --include-info-json s=s/target/doc.parts --extern t=t/target/libt.rmeta --extern s=s/target/libs.rmeta -L t/target i/src/lib.rs
```

> Merge the docs with `cp`. This can be avoided if `--out-dir=$(MERGED)` is used for all of the rustdoc calls. We copy here to illustrate that documenting `s` is independent of documenting `t`, and could happen on separate machines.

Author: `cp` is the stand-in for the step that takes docs from external crates and collects them in a single web root. Depending on the location of the external crates, this step may look very different:
* It may do nothing; `Cargo` generates the docs in a single `doc/` folder.
* `cp` or `mv` files from one location on the filesystem to another. This is shown in the example in the RFC.
* `scp`, `rsync`, `wget --recursive`, a proprietary protocol, etc. This RFC is intended to enable CCI in crates that have been documented on separate machines.

Because of the wide variety of locations that external docs may be, I believe that the build system is best positioned to collect them into a single location.

When you pass the `--merge=none` flag to document the external crates, the shared resources are not generated. Only `crate-info.json` and the item docs will be generated for these crates. The shared resources are only generated when you pass `--merge=auto` (default) or `--merge=write-only`, which is only provided when documenting the index or root crate. There are not any redundant files to bloat the disk space.

> Merge the docs with `cp`. This can be avoided if `--out-dir=$(MERGED)` is used for all of the rustdoc calls. We copy here to illustrate that documenting `s` is independent of documenting `t`, and could happen on separate machines.

Reviewer 5: Then please update so that it is mentioned that rustdoc will handle the copying internally as we don't want to rely on external tools.

> Merge the docs with `cp`. This can be avoided if `--out-dir=$(MERGED)` is used for all of the rustdoc calls. We copy here to illustrate that documenting `s` is independent of documenting `t`, and could happen on separate machines.

Author: Okay, I will update it so that rustdoc does the copying.

│       │   ├── SourceSerif4-Regular-46f98efaafac5295.ttf.woff2
│       │   ├── src-script-e66d777a5a92e9b2.js
│       │   └── storage-118b08c4c78b968e.js
│       └── t
│           ├── all.html
│           ├── index.html
│           ├── sidebar-items.js
│           └── trait.T.html
├── s
│   ├── src
│   │   └── lib.rs
│   └── target
│       ├── doc
│       │   ├── help.html
│       │   ├── .lock
│       │   ├── s
│       │   │   ├── all.html
│       │   │   ├── index.html
│       │   │   ├── sidebar-items.js
│       │   │   └── struct.S.html
│       │   ├── settings.html
│       │   └── src
│       │       └── s
│       │           └── lib.rs.html
│       ├── doc.parts
│       │   └── s
│       │       └── crate-info.json
│       └── libs.rmeta
└── t
    ├── src
    │   └── lib.rs
    └── target
        ├── doc
        │   ├── help.html
        │   ├── .lock
        │   ├── t
        │   │   ├── all.html
        │   │   ├── index.html
        │   │   ├── sidebar-items.js
        │   │   └── trait.T.html
        │   ├── settings.html
        │   └── src
        │       └── t
        │           └── lib.rs.html
        ├── doc.parts
        │   └── t

> │       │       └── crate-info.json

Reviewer 5: Sorry it bugs me. :laughing: 

```suggestion
        │       └── crate-info.json
```

## New flag: `--merge=auto|none|write-only`

This flag controls two internal paramaters: `read_rendered_cci`, and `write_rendered_cci`.

When `write_rendered_cci` is active, rustdoc will output the rendered parts to the doc root (`--out-dir`). Rustdoc will generate files like `doc/search-index.js`, `doc/search.desc`, `doc/index.html`, etc if and only if this parameter is true.

When `read_rendered_cci` is active, rustdoc will look in the `--out-dir` for rendered cross-crate info files. These files will be used as the base. Any new parts that rustdoc generates with its current invocation and any parts fetched with `include-info-json` will be appended to these base files. When it is disabled, the cross-crate info files start empty and are populated with the current crate's info and any crates fetched with `--include-info-json`.

> * `--merge=auto` (`read_rendered_cci && write_rendered_cci`) is the default, and reflects the current behavior of rustdoc. 

Reviewer 5: I'm not a big fan of `auto`. Why not `read-write` instead? Would be more explicit (took me a while to get why it was different to `write-only` :sweat_smile:).

Author: That sounds good!

This proposal is capable of addressing two primary use cases. It allows developers to enable CCI in these scenarios:
* Documenting a crate and its transitive dependencies in parallel in build systems that require build actions to be independent
* Producing a documentation index of a large number of crates, in such a way that if one crate is updated, only the updated crates and an index have to be redocumented. This scenario is demonstrated in the Guide-level explanation.

CCI is not automatically enabled in either situation. A combination of the `--include-info-json`, `--merge`, and `--write-info-json` flags are needed to produce this behavior. This RFC provides a minimal set of tools that allow developers of build systems, like Bazel and Buck2, to create rules for these scenarios.

Discussion of whether additional features should be included to facilitate this merge step can be found in Unresolved questions (Index crate).

## Compatibility

This RFC does not alter previous compatibility guarantees made about the output of rustdoc. In particular it does not stabilize the presence of the rendered cross-crate information files, their content, or the HTML generated by rustdoc.

> In the same way that the [rustdoc HTML output is unstable](https://rust-lang.github.io/rfcs/2963-rustdoc-json.html#:~:text=The%20HTML%20output%20of%20rustdoc,into%20a%20different%20format%20impractical), the contents of the `crate-info.json` will be considered unstable. Between versions of rustdoc, breaking changes to the content of `crate-info.json` should be expected. Only the presence of a `crate-info.json` file is promised, under `--write-info-json`. Merging cross-crate information generated by disparate versions of rustdoc is not supported.

Reviewer 5: ```suggestion
In the same way that the [rustdoc HTML output is unstable](https://rust-lang.github.io/rfcs/2963-rustdoc-json.html#:~:text=The%20HTML%20output%20of%20rustdoc,into%20a%20different%20format%20impractical), the content of the `crate-info.json` will be considered unstable. Between versions of rustdoc, breaking changes to the content of `crate-info.json` should be expected. Only the presence of a `crate-info.json` file is promised, under `--write-info-json`. Merging cross-crate information generated by disparate versions of rustdoc is not supported.
```

Reviewer 5: You say:

> Merging cross-crate information generated by disparate versions of rustdoc is not supported.

But in the description of the content of the file, you don't mention that the version of rustdoc is stored or how it's stored.

Author: Do you have a suggestion about how to version the `crate-info.json` files? We could put a semver in file and accept an `--info-json-version` flag Rustdoc would fail if it is passed an `--include-info-json` with an incompatible file.

Reviewer 5: That's a tricky question. Currently the rustdoc JSON output has a simple counter but it's something we're thinking about changing, but we didn't find a new versioning system yet (you can read about it [here](https://rust-lang.zulipchat.com/#narrow/stream/266220-t-rustdoc/topic/Stabilising.20JSON.20output.3F)).

So really not sure what to pick, but this is definitely a blocker for this RFC.

Author: I think that an incremental single-digit version number makes the most sense. The RFC only discusses the case where crate-info.json files are written by rustdoc, and consumed by the same version rustdoc, so it doesn't seem important to detail what constitutes a breaking change. Also the content of the files is tied to the internal details of rustdoc (including the details of the search index). I'll make rustdoc write an incremental single-digit version to the crate-info.json file, and rustdoc will fail if it's called with an `--include-info-json` pointing to a `crate-info.json` with a version that does not match its internal version.

The implementation of the RFC itself is designed to produce only minimal changes to cross-crate info files and the HTML output of rustdoc. Exhaustively, the implementation is allowed to 
* Change the sorting order of trait implementations, type implementations, and other cross-crate info in the HTML output of rustdoc

> * Add a comment on the last line of generated HTML pages, to store metadata relevant to appending items to them

Reviewer 5: I'm not sure to see the need for this requirement. Do you have an example in mind?

Author: I don't want to make any changes to the existing tools that use rustdoc across multiple crates by documenting every crate in the same output directory. If we stop supporting this mode, there are a lot of tools that we would break, including the rustdoc test suite, Cargo, and numerous other proprietary tools. That's why the RFC and my implementation has tighter requirements on what it's going to do, and that's why it has a default `--mode=read-write`.

I expect the sorting order of cross-crate information to change because the existing sorting order depends on specific implementation details, like whether we're sorting by the serialized or native representations of the CCI.

I am allowing adding a comment to the last line of the rendered files, because this is how my implementation implements `read_rendered_cci`. It finds out where to insert the next part into an existing file by looking at a comment that is written to the last line.

Reviewer 5: Hum. There is already [`--html-after-content`](https://doc.rust-lang.org/rustdoc/command-line-arguments.html?#--html-after-content-include-more-html-after-the-content). Wouldn't it cover this use case?

<https://github.com/facebook/buck2/blob/d390e31632d3b4a726aaa2aaf3c9f1f63935e069/prelude/rust/build.bzl#L213>

## Ninja [(GN)](https://fuchsia.dev/fuchsia-src/development/build/build_system/intro) + Fuchsia

Currently, the Fuchsia project runs rustdoc on all of their crates to generate a [documentation index](https://fuchsia-docs.firebaseapp.com/rust/rustdoc_index/). This index is effectively generated as an [atomic step](https://cs.opensource.google/fuchsia/fuchsia/+/main:tools/devshell/contrib/lib/rust/rustdoc.py) in the build system. It takes [3 hours](https://ci.chromium.org/ui/p/fuchsia/builders/global.ci/firebase-docs/b8744777376580022225/overview) to document the ~2700 crates in the environment. With this proposal, building each crate's documentation could be done as separate build actions, which would have a number of benefits. These include parallelism, caching (avoid rebuilding docs unnecessarily), and robustness (automatically reject pull requests that break documentation).

# Unresolved questions

## Index crate?

> Require users to generate documentation bundles via an index crate (current) vs. creating a new mode to allow rustdoc to run without a target crate (proposed)

Reviewer 5: Missing a dot at the end of the sentence.

The proposition, to allow users of rustdoc the flexibility of not having to produce an index, is to allow rustdoc to be run in a mode where no target crate is provided. It would generate rendered cross-crate information based only on what is provided through `--include-info-json`. The source crates are provided to rustdoc, through a mechanism like `--extern`, rustdoc merges and writes the CCI, and copies the item and module links to the doc root. This would require more extensive changes, as rustdoc assumes that it is invoked with a target crate. This mode is somewhat analogous to the [example scraping mode](https://github.com/rust-lang/rfcs/blob/master/text/3123-rustdoc-scrape-examples.md). Having to create an index crate, that actively uses all of the crates in the environment, might prohibit the use of this feature in settings where users do not intend to produce an index, or where exhaustively listing all dependencies (to `--extern` them) is difficult.

## Unconditionally generating the `doc.parts` files?

> Generate no extra files (current) vs. unconditionally creating `doc.parts` to enable more complex future CCI (should consider)

Reviewer 5: Missing a dot at the end of the sentence.

This proposal proposes to continue reading from the rendered cross-crate information under the default `--merge=auto`. It can also read `crate-info.json` files, under `--include-info-json`. However, there are several issues with reading from the rendered CCI that must be stated:
* Every rustdoc process outputs the CCI to the same doc root by default
* It is difficult to extract the items in a diverse set of rendered HTML files. This is anticipating of the CCI to include HTML files that, for example, statically include type+trait implementations directly
* Reading exclusively from `crate-info.json` is simpler than the existing `serde_json` dependency for extracting the blobs, as opposed to handwritten CCI-type specific parsing (current)
* With this proposal, there will be duplicate logic to read from both `doc.parts` files and rendered CCI.

[@jsha proposes](https://github.com/rust-lang/rfcs/pull/3662#issuecomment-2184077829) unconditionally generating and reading from `crate-info.json`, with no appending to the rendered crate info.

## Item links?

> Require users to pass `--extern-html-root-url` on all external dependencies (current), vs. add a new flag to facilitate missing docs links being generated across all user crates (should consider)

Reviewer 5: Missing dot.

Cross-crate links are an important consideration when merging documentation. rustdoc provides a flag, `--extern-html-root-url`, which can provide fine-grain control over the generated URL for documentation. We believe this is already sufficient for generating cross-crate links, and we will make no changes to the generation of item links. Example usage of this flag is in the Guide-level explanation.

A proposal is to provide more options to facilitate cross-crate links. For example, we could add a flag that implies `--extern-html-root-url` and `--include-info-json` on all crates passed through `--extern`. It could be something like `--extern-html-root-url-all-user-crates`. User crates would be taken to mean the transitive dependencies of the current crate, excluding the standard library. This is complicated by things like <https://doc.rust-lang.org/cargo/reference/unstable.html#build-std>.

## Reuse existing option?

> Create a new flag, `--merge` (proposed), vs. use existing option `no_emit_shared`

Reviewer 5: Missing dot.

There is a render option, `no_emit_shared`, which is used to conditionally generate the cross-crate information. It also controls the generation of static files, user CSS, etc.

This option is not configurable from the command line, and appears to be enabled unless rustdoc is run in its example scraping mode.

We could make it configurable from the command line, unconditionally generate `doc.parts`, and use it to gate the merging of CCI.

We could also make it configurable from the command line, and use it to gate the generation of `doc.parts` and the generation of all of the shared files.

We could also leave it as-is: always false unless we're scraping examples, and gate the generation of `doc.parts` and the generation of all of the shared files.

# Future possibilities

This change could begin to facilitate type aliases and the trait aliases being
statically compiled as part of the .html documentation, instead of being loaded
as separate JavaScript files. Each type and trait alias could be stored as an
HTML part, which are then merged into the regular documentation.

> Another possibility is for `doc.parts` to be distributed on `docs.rs` along with the regular documentation. This would facilitate a mode where documentation of the dependencies could be downloaded externally, instead of being rebuilt locally.

Reviewer 5: That would be very interesting!

Author: That is true. It would really only be useful if you wanted to avoid the additional work of documenting a crate locally, but the rmeta and the generated docs were available externally.

Reviewer 5: My after-thought point here was that distributing these files on docs.rs would likely not bring any advantage (except for the search index maybe).

# Reference-level explanation

The existing cross-crate information files, like `search-index.js`, all are lists of elements, rendered in an specified way (e.g. as a JavaScript file with a JSON array or an HTML index page containing an unordered list). The current rustdoc (in `write_shared`) pushes the current crate's version of the CCI into the one that is already found in `doc`, and renders a new version. The rest of the proposal uses the term **part** to refer to the pre-merged, pre-rendered element of the CCI. This proposal does not add any new CCI or change their contents (modulo sorting order, whitespace).

## New file: `crate-info.json`

The `crate-info.json` file contains the unmerged contents of a single crates' version of their corresponding CCI. This file is written if the flag `--write-info-json=path/to/crate-info.json` is provided.

`crate-info.json` is a JSON object with several key-value pairs. Every value is a JSON array. Every element of the JSON array is a two-element array: a destination filename relative to the doc root, and the representation of the part. The representation of the part depends on the type of CCI that it describes.

* key: `version`

> The value will be a string encoding of a version number. If rustdoc is provided a `--include-info-json` flag that points to a `crate-info.json` file with a version number that it cannot support, it will fail. Rustdoc only guarantees that it will accept `crate-info.json` files written by the same version of rustdoc. There is no forward or backward compatibility. A single number prefixed with capital V is chosen for convenience, as `crate-info.json` does not make distinctions between breaking and non-breaking changes. Rustdoc is the only explicitly supported consumer of `crate-info.json`.

Reviewer 5: Thanks for writing the section. Now we just need the rustdoc team to agree on a versioning system (because this is the one we use in the JSON output and we're discussing about changing it, so better fix it before an RFC to use it is approved ^^').

Reviewer 5: No but if we could at least have the version format to be definitive, that'd make our (rustdoc) work much simpler if we can rely on this key whatever the rest of the file looks like.

Author: I agree that the content of `crate-info.json` is opaque. I think that how we determine whether rustdoc is pointed to an incompatible `crate-info.json`, and how we report that error message, is an implementation detail. The version number, just like the content of the file, will not be visible or consumable by users of rustdoc.

Reviewer 5: Hum... That's a fair point. Ok, then I'm ok with this way of doing it. Then please add this explanation in the RFC text and remove the part about what the version is about, we'll discuss it in the implementation then.

Author: Okay, I removed many of the specific details about the contents of crate-info.json, including the version key, and explained that they were unstable.

In this mode, rendered HTML and *finalized* cross-crate information are generated into a `target/doc/my-final-crate` folder. No *incremental* parts are generated (i.e., no `target/doc.parts/my-final-crate`).

When a user documents the final crate, they will provide  `--include-parts-dir=<crate-specific path selected previously>`, `--include-rendered-docs=<out dirs chosen previously>` for each one of the dependencies, and `--merge=finalize`. They will provide `--extern-html-root-url`, in the way described in 'Document an intermediate crate'.

# Reference-level explanation

The existing cross-crate information files, like `search-index.js`, all are lists of elements, rendered in an specified way (e.g. as a JavaScript file with a JSON array or an HTML index page containing an unordered list). The current rustdoc (in `write_shared`) pushes the current crate's version of the CCI into the one that is already found in `doc`, and renders a new version. The rest of the proposal uses the term **part** to refer to the pre-merged, pre-rendered element of the CCI. This proposal does not add any new CCI or change their contents (modulo sorting order, whitespace).

> ## New directory: `doc.parts/<crate-name>/crate-info.json`

Reviewer 2: This seems to suggest that `doc.parts` will always be the name of the directory. It'd be good to be very explicit that this is just a suggestion for the name; rustdoc will use the name provided by users, with no default name (unlike `doc`).

This flag is similar to `--extern-html-root-url` in that it only needs to be provided for externally documented crates. The flag `--extern-html-root-url` controls hyperlink generation. The hyperlink provided in `--extern-html-root-url` never accessed by rustdoc, and represents the final destination of the documentation. The new flag `--include-parts-dir` tells rustdoc where to search for the `doc.parts` directory at documentation-time. It must not be a URL.

In the Guide-level explanation, for example, crate `i` needs to identify the location of `s`'s parts. Since they could be located in an arbitrary directory, `i` must be instructed on where to fetch them. In this example, `s`'s parts happen to be in `./s/target/doc.parts/s`, so rustdoc is called with `--include-parts-dir=s/target/doc.parts/s`.

## New flag: `--include-rendered-docs=<path/to/target/doc/extern-crate-name>`

> Rustdoc will assume that `<path/to/target/doc>` was used as the `--out-dir` for `<extern-crate-name>`. This documentation will be copied into the directory specified by `--out-dir`. Rustdoc will effectively run `cp -r <path/to/target/doc/extern-crate-name> <current --out-dir>`.

Reviewer 2: I'd really prefer not to add it. As you yourself say, users will often have more complex needs for copying the docs, and there's no advantage other a tiny convenience for rustdoc to include this functionality. The only people using these flags will be implementers of build systems or users who have very complex build requirements. Also, they'll be manually controlling the location and structure of the build directories. So I don't see the downside in requiring them to use a simple `cp` here. If we include the feature, there's also the risk for feature creep where users will want us to add more complex functionality over time.

Note that unlike `--include-parts-dir`, there's no rustdoc-specific behavior. That flag needs to exist because it affects rustdoc's behavior and knowledge of the workspace.

Author: I think these are good arguments! @GuillaumeGomez had a valid concern about relying on external tools. However, even without `--include-rendered-docs`, you would only need to rely on external tools for merging in the case of multiple `--out-dir`s, which is an extreme scenario. Having multiple `--out-dir`s is not necessary to avoid having multiple build actions target the same files, since rustdoc writes item docs under a separate directory anyway. And this RFC is mostly about providing options to avoid appending to files within a single `--out-dir`. Even though the multiple `--out-dir` case is anticipated by this RFC, if someone is deciding to split their rustdoc builds in arbitrary ways they should expect to do some manual work.

```shell
rustdoc -Z unstable-options --crate-name=i --crate-type=lib --edition=2021  --enable-index-page --out-dir=i/target/doc --extern-html-root-url s=$(MERGED) --extern-html-root-url t=$(MERGED) --extern-html-root-url i=$(MERGED) --read-rendered-cci=false --fetch-parts t=t/target/doc.parts --fetch-parts s=s/target/doc.parts --extern t=t/target/libt.rmeta --extern s=s/target/libs.rmeta -L t/target i/src/lib.rs
```

Merge the docs with `cp`. This can be avoided if `--out-dir=$(MERGED)` is used for all of the rustdoc calls. We copy here to illustrate that documenting `s` is independent of documenting `t`, and could happen on separate machines.

```shell
cp -r s/target/doc/* t/target/doc/* i/target/doc/* merged/doc
```

Browse `merged/doc/index.html` with cross-crate information.

In general, instead of two crates in the environment (`s` and `t`) you could have thousands. Upon any changes, only the index and the crates that are changed have to be re-documented.

<details>
<summary>Click here for a directory listing after running the example above.</summary>

<pre>
$ tree . -a
.
├── i
│   ├── Cargo.lock
│   ├── Cargo.toml

> │   ├── .gitignore

Author: Just manually deleted these files; I will re-generate the `tree` when WIP is ready for that

## Suggested workflows

With this proposal, there are three modes of invoking rustdoc: `--merge=shared`, `--merge=none`, and `--merge=finalize`.

### Default workflow: mutate shared directory: `--merge=shared`

In this workflow, we document a single crate, or a collection of crates into a shared output directory that is continuously updated.
Files in this output directory are modified by multiple rustdoc invocations. Use `--merge=shared`, and specify the same `--out-dir` to every invocation of rustdoc. `--merge=shared` will be the default value if `--merge` is not provided. This is the workflow that Cargo uses, and only mode of invoking rustdoc before this RFC.

### Document intermediate crates: `--merge=none`

Document non-root, non-index crates using a dedicated HTML output directory and a dedicated "parts" output directory. No cross-crate data nor rendered HTML output is included from other crates.

This mode only renders the HTML item documentation for the current crate. It does not produce a search index, cross-crate trait implementations, or an index page. It is expected that users follow this mode with 'Document a final crate' if these cross-crate features are desired.

> In this mode, a user may specify a different `--out-dir` to every invocation of rustdoc. Additionally, a user will provide `--parts-out-dir=<path to crate-specific directory>` and `--merge=none` when documenting every crate.

Author: In many contexts, its okay to run multiple invocations of rustdoc concurrently with the same out dir. Most of the files do not overlap (the item docs are separate). The files that are overwritten by multiple rustdoc invocations are written inside [`write_shared`](https://github.com/EtomicBomb/rust/blob/rfc/src/librustdoc/html/render/write_shared.rs#L69), which acquires an flock. In my understanding, `cargo doc` uses this when it launches multiple rustdoc processes at once.

* `--include-parts-dir=path/to/doc.parts/<crate-name>`: Include cross-crate information from this previously written `doc.parts` directories into a collection that will be written by the current invocation of rustdoc. May only be provided with `--merge=finalize`. May be provided any number of times.
* `--merge=none`: Do not write cross-crate information to the `--out-dir`. The flag `--parts-out-dir` may instead be provided with the destination of the current crate's cross-crate information parts.
* `--merge=shared` (default): Append information from the current crate to any info files found in the `--out-dir`.

> * `--merge=finalize`: Write cross-crate information from the current crate and any crates included via `--include-parts-dir` to the `--out-dir`, overwriting conflicting files. This flag may be used with or without an input crate root, as it can links existing docs.

Reviewer 3: ```suggestion
* `--merge=finalize`: Write cross-crate information from the current crate and any crates included via `--include-parts-dir` to the `--out-dir`, overwriting conflicting files. This flag may be used with or without an input crate root, as it can link existing docs.
```
